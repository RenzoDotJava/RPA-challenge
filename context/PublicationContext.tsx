import { createContext } from "react";
import { IUserPublications } from "../interfaces/publication.interface";

type PublicationContextProps = {
	filteredUsersPublications: IUserPublications[];
	search: string;
	loading: boolean;
	getUsersPublications: (refresh: boolean) => Promise<IUserPublications[]>;
	searchFilter: (text: string) => void;
	onRefreshPublications: () => void;
};

export const PublicationContext = createContext<PublicationContextProps>(
	{} as PublicationContextProps
);
