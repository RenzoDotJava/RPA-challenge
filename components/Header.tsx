import { useContext } from "react";
import { View, TextInput, StyleSheet, Platform } from "react-native";
import { AntDesign } from "@expo/vector-icons";
import { PublicationContext } from "../context/PublicationContext";

const Header: React.FC = () => {
	const { search, searchFilter } = useContext(PublicationContext);

	return (
		<View style={styles.header}>
			<View style={styles.container}>
				<TextInput
					style={styles.searchInput}
					onChangeText={(text) => searchFilter(text)}
					value={search}
				/>
				<AntDesign name="search1" size={20} color="white" />
			</View>
		</View>
	);
};

const styles = StyleSheet.create({
	header: {
		flexDirection: "row",
		alignItems: "center",
		paddingBottom: 8,
		backgroundColor: "black",
	},
	container: {
		flexDirection: "row",
		marginHorizontal: 12,
		marginTop: 10,
		marginBottom: 1,
		borderBottomColor: "white",
		borderBottomWidth: 2,
	},
	searchInput: {
		flex: 1,
		color: "white",
		paddingBottom: Platform.OS === "android" ? 4 : 15,
	},
});

export default Header;
