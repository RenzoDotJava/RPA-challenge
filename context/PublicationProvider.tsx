import AsyncStorage from "@react-native-async-storage/async-storage";
import { useState, useEffect, useCallback } from "react";
import { IPublication, IUserPublications } from "../interfaces/publication.interface";
import { PublicationContext } from "./PublicationContext";
import { PublicationService } from "../services/publicationService";

export const PublicationProvider: React.FC<any> = ({ children }) => {
	const [usersPublications, setUsersPublications] = useState<IUserPublications[]>([]);
	const [filteredUsersPublications, setFilteredUsersPublications] = useState<IUserPublications[]>([]);
  const [search, setSearch] = useState<string>("");
  const [loading, setLoading] = useState<boolean>(false);

  const publicationService: PublicationService = new PublicationService();

  useEffect(() => {
		let isCanceled: boolean = false;

		getUsersPublications().then((res: IUserPublications[]) => {
			if (isCanceled) return;

			setUsersPublications(res);
			setFilteredUsersPublications(res);
		});

		return () => {
			isCanceled = true;
		};
	}, []);

  const groupPublicationsByUserId = (
    publications: IPublication[]
  ): IUserPublications[] => {
    return publications.reduce((acc: IUserPublications[], val: IPublication) => {
      let publication: IPublication = {
        id: val.id,
        title: val.title,
        body: val.body.replace(/\n|\r/g, ""),
      };
  
      const existingUser: IUserPublications | undefined = acc.find(
        (group: IUserPublications) => group.userId === val.userId
      );
  
      if (existingUser) {
        existingUser.data.push(publication);
      } else {
        acc.push({
          userId: val.userId,
          data: [publication],
        });
      }
  
      return acc;
    }, []);
  };

  const getPublicationsByUsersAmount = (
    amount: number = 5
  ): Promise<IUserPublications[]> => {
    let indexes: number[] = Array.from({ length: amount }, (val, i) => i);
    return Promise.all(
      indexes.map((index) => {
        return publicationService.getPublicationsByUserId(index + 1);
      })
    ).then(async (res: IPublication[][]) => {
      const groupedPublications: IUserPublications[] =
        groupPublicationsByUserId(res.flat());
  
      await AsyncStorage.setItem(
        "usersPublications",
        JSON.stringify(groupedPublications)
      );
  
      return groupedPublications;
    });
  };

  const getUsersPublications = async (
    refresh: boolean = false
  ): Promise<IUserPublications[]> => {
    let json: string | null = "";
    let usersPublications: IUserPublications[] = [];
  
    if (!refresh) json = await AsyncStorage.getItem("usersPublications");
  
    if (!json || json == "") {
      await getPublicationsByUsersAmount().then((res) => {
        usersPublications = res;
      });
    } else {
      usersPublications = JSON.parse(json);
    }
  
    return usersPublications;
  };

  const searchFilter = (text: string) => {
		if (text.trim() == "") {
			setFilteredUsersPublications(usersPublications);
		} else {
			const newPublications: IUserPublications[] = usersPublications
				.filter((item) =>
					item.data.some((d) => d.title.includes(text.toLowerCase()))
				)
				.map((item) => ({
					userId: item.userId,
					data: item.data.filter((d) =>
						d.title.includes(text.toLowerCase())
					),
				}));

			setFilteredUsersPublications(newPublications);
		}

		setSearch(text);
	};

  const onRefreshPublications = useCallback(() => {
		setLoading(true);
		getUsersPublications(true)
			.then((res: IUserPublications[]) => {
				setUsersPublications(res);
				setFilteredUsersPublications(res);
			})
			.catch(() => {
				setUsersPublications([]);
				setFilteredUsersPublications([]);
			})
			.finally(() => {
				setSearch("");
				setLoading(false);
			});
	}, []);

	return (
		<PublicationContext.Provider
			value={{
				filteredUsersPublications,
        search,
        loading,
        getUsersPublications,
        searchFilter,
        onRefreshPublications
			}}
		>
			{children}
		</PublicationContext.Provider>
	);
};
