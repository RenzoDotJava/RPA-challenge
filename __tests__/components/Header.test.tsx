import { render, screen, fireEvent } from "@testing-library/react-native";
import Header from "../../components/Header";
import { PublicationContext } from "../../context/PublicationContext";

const userPublications = [
	{
		userId: 1,
		data: [
			{
				id: 1,
				title: "title mockup 1",
				body: "body mockup 1",
			},
		],
	},
	{
		userId: 2,
		data: [
			{
				id: 2,
				title: "title mockup 2",
				body: "body mockup 2",
			},
		],
	},
];

describe("Testing <Header />", () => {
	beforeEach(() => jest.clearAllMocks());

	it("Matching snapshot", () => {
		render(
			<PublicationContext.Provider
				value={{
					filteredUsersPublications: [],
					search: "",
					loading: false,
					getUsersPublications: jest.fn(),
					searchFilter: jest.fn(),
					onRefreshPublications: jest.fn()
				}}
			>
				<Header />
			</PublicationContext.Provider>
		);

		expect(screen).toMatchSnapshot();
	});

	it("Running searchFilter function when changing the text", () => {
		const searchFilterMockup = jest.fn();
		const text = "title";

		render(
			<PublicationContext.Provider
				value={{
					filteredUsersPublications: userPublications,
					search: "",
					loading: false,
					getUsersPublications: jest.fn(),
					searchFilter: searchFilterMockup,
					onRefreshPublications: jest.fn()
				}}
			>
				<Header />
			</PublicationContext.Provider>
		);

		const textInput = screen.getByTestId("search-bar");

		fireEvent.changeText(textInput, text);

		expect(searchFilterMockup).toHaveBeenCalledWith(text);
	});

});
