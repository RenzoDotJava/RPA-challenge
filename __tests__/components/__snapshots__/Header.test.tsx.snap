// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing <Header /> Matching snapshot 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "backgroundColor": "black",
      "flexDirection": "row",
      "paddingBottom": 8,
    }
  }
  testID="header"
>
  <View
    style={
      {
        "borderBottomColor": "white",
        "borderBottomWidth": 2,
        "flexDirection": "row",
        "marginBottom": 1,
        "marginHorizontal": 12,
        "marginTop": 10,
      }
    }
  >
    <TextInput
      onChangeText={[Function]}
      style={
        {
          "color": "white",
          "flex": 1,
          "paddingBottom": 15,
        }
      }
      testID="search-bar"
      value=""
    />
    <Text />
  </View>
</View>
`;
